src/filter/kuadrant_filter.rs

```diff
@@ -0,0 +1,187 @@
+use crate::action_set_index::ActionSetIndex;
+use crate::filter::operations::{
+    GrpcMessageReceiverOperation, GrpcMessageSenderOperation, HeadersOperation, Operation,
+};
+use crate::runtime_action_set::RuntimeActionSet;
+use crate::service::{GrpcErrResponse, GrpcRequest, HeaderResolver};
+use log::{debug, warn};
+use proxy_wasm::traits::{Context, HttpContext};
+use proxy_wasm::types::{Action, Status};
+use std::mem;
+use std::rc::Rc;
+
+pub(crate) struct KuadrantFilter {
+    context_id: u32,
+    index: Rc<ActionSetIndex>,
+    header_resolver: Rc<HeaderResolver>,
+
+    grpc_message_receiver_operation: Option<GrpcMessageReceiverOperation>,
+    headers_operations: Vec<HeadersOperation>,
+}
+
+impl Context for KuadrantFilter {
+    fn on_grpc_call_response(&mut self, token_id: u32, status_code: u32, resp_size: usize) {
+        debug!(
+            "#{} on_grpc_call_response: received gRPC call response: token: {token_id}, status: {status_code}",
+            self.context_id
+        );
+        let receiver = mem::take(&mut self.grpc_message_receiver_operation)
+            .expect("We need an operation pending a gRPC response");
+
+        let mut ops = Vec::new();
+
+        if status_code != Status::Ok as u32 {
+            ops.push(receiver.fail());
+        } else if let Some(response_body) = self.get_grpc_call_response_body(0, resp_size) {
+            ops.extend(receiver.digest_grpc_response(&response_body));
```

Comment:
So we need a vec because when digesting the grpc response we might need to add an extra operation (AddHeaders) before sending the next gprc request, right?

Reply:
Yep exactly, right now this can return two ops and we process them in sequence, based on how this all currently works. However this was more so to open the seam to allow us to change how the stuff beneath it works - for example what if we have 5 actions that just want to add headers before continuing? we can just build a vec and let the filter process them all until it needs to 'wait'